### PR TITLE
MINIFI-30 added changes to the config transformer to support multiple…

### PR DESCRIPTION
**Pull request by JPercivall**
… processors and exfiling using a processor

### CODE REVIEW

**Comment by apiri** [comment]
starting review

File: minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java [hunk 1]
```diff
@@ -219,6 +223,7 @@ private static void writeNiFiProperties(Map<String, Object> topLevelYaml, Output
             Map<String, Object> componentStatusRepo = (Map<String, Object>) topLevelYaml.get(COMPONENT_STATUS_REPO_KEY);
             Map<String, Object> securityProperties = (Map<String, Object>) topLevelYaml.get(SECURITY_PROPS_KEY);
             Map<String, Object> sensitiveProperties = (Map<String, Object>) securityProperties.get(SENSITIVE_PROPS_KEY);
+            Map<String, Object> provenanceProperties = (Map<String, Object>) securityProperties.get(PROVENANCE_REPORTING_KEY);
```
Comment (apiri) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: Believe this should be topLevelYaml and not security properties.

Comment (apiri) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: Upon further inspection I see this is unused and the provenance properties are gathered in createFlowXml.  Likely just remove unless you would prefer to gather these maps up front.

Comment (JPercivall) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: Will remove, I think I originally thought something in the nifi.properties needed it but I guess I was mistaken

File: minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java [hunk 2]
```diff
@@ -428,13 +436,26 @@ private static void addProcessGroup(final Element parentElement, Map<String, Obj
...
+            if (processors != null) {
+                for (Object processorConfig : processors) {
+                    addProcessor(element, (Map<String, Object>) processorConfig);
+                }
+            }
 
-            Map<String, Object> remoteProcessingGroup = (Map<String, Object>) topLevelYaml.get(REMOTE_PROCESSING_GROUP_KEY);
-            addRemoteProcessGroup(element, remoteProcessingGroup);
+            Map<String, Object> exfilComponent = (Map<String, Object>) topLevelYaml.get(EXFIL_COMPONENT_KEY);
+            if ("Site-To-Site".equals(getValueString(exfilComponent, TYPE_KEY))) {
```
Comment (apiri) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: minor:  would be nice to make this a constant as it is used in two different places for checks

File: minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java [hunk 3]
```diff
@@ -428,13 +436,26 @@ private static void addProcessGroup(final Element parentElement, Map<String, Obj
...
+                addRemoteProcessGroup(element, exfilComponent);
+            } else if ("PROCESSOR".equals(getValueString(exfilComponent, TYPE_KEY))) {
```
Comment (apiri) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: minor:  would be nice to make this a constant as it is used in a few different places for the same purpose

File: minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java [hunk 4]
```diff
@@ -616,15 +630,21 @@ private static void addConnection(final Element parentElement, Map<String, Objec
             addTextElement(element, "labelIndex", "1");
             addTextElement(element, "zIndex", "0");
 
-            addTextElement(element, "sourceId", "Processor");
+            addTextElement(element, "sourceId", getValueString(connectionProperties, SOURCE_NAME_KEY));
```
Comment (apiri) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: For properties like these that are required, we should have an indication that it is needed and an associated exception if it is not present.

**Comment by apiri** [comment]
Overall looks like a good start to accomplish what is needed for this effort.  Some additional validation is needed to enforce the required properties such that we can catch them before attempting to begin the flow anew after a transformation.

File: minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java (hunk 4 above)
Comment (apiri) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: One additional thought I had was to have another class that perhaps does the value checking and enforces the config YAML and then passes a value object (a collection of all the backing maps) to the ConfigTransformer.  Long term, we may not have this particular process and could isolate the actual validation and checking elsewhere.  In lieu of the manual checks, we may be able to make use of something like [Kwalify](http://www.kuwata-lab.com/kwalify/) (no clue as to its quality) which might help in cutting down on the onerous and manual if statements for each individual scalar.

Comment (JPercivall) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/ConfigTransformer.java]: That is a very good way forward and I will refactor this to be two classes. I'll look in to Kwalify and google a bit to see what others are using for YAML schema validation.

**Comment by JPercivall** [comment]
@apiri just pushed out a new commit refactoring the validation to a separate classes. There are merge conflicts with master in ConfigTransformer that I will address tomorrow.

**Comment by JPercivall** [comment]
@apiri rebased and pushed

**Comment by JPercivall** [comment]
Thinking a bit more, I don't see a reason to not allow multiple RPGs either. Will make the adjustment

**Comment by JPercivall** [comment]
@apiri added support for multiple RPGs and multiple Input Ports within those RPGs

**Comment by apiri** [comment]
This looks good overall.  I would avoid the sheer ClassCast exception handling and prefer to do instanceof checks instead.  Reads a bit cleaner.

Also, for the majority of cases, we have a pattern that is roughly:
```Java
  value getKeyAsType(valueMap, key, target, wrapperName, reason, optional, defaultValue) {
        value = valueMap.get(key);
        if (value == null) {
            if (optional && defaultValue != null) {
                return defaultValue;
            }
        } else {
            targetClass = target.getClass()
            if (targetClass.isInstance(value)) {
                return value;
            } else {
                addValidationIssue(key, wrapperName, reason + targetClass);
            }
        }
        return null;
    }
```

This with a few overloaded signatures would allow cutting down on a significant number of lines and centralize logic throughout.  I saw a few special cases, which should likely be left to their own, but it looks like we can standardize the bulk of the validation.

**Comment by JPercivall** [comment]
@apiri just pushed out a new commit centralizing the logic for getting the various values

File: minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/common/BaseSchema.java [hunk 5]
```diff
@@ -51,18 +55,80 @@ public void addValidationIssue(String keyName, String wrapperName, String reason
...
-    public void addIssuesIfNotNull(BaseSchema baseSchema) {
-        if (baseSchema != null) {
-            validationIssues.addAll(baseSchema.getValidationIssues());
+
+    public <T> T getMapAsType(Map valueMap, String key, Class targetClass, String wrapperName, boolean required) {
+        Object obj = valueMap.get(key);
+        return interpretValueAsType(obj, key, targetClass, wrapperName, required);
+    }
+
+    public void instantiateList(List list, String simpleListType, Class targetClass, String wrapperName){
```
Comment (apiri) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/common/BaseSchema.java]: Is it fair to say that this is more of a transform/mapping to type for the already existing list?

**Comment by apiri** [comment]
Hey @JPercivall, changes look great and overall much cleaner.  Just had the question regarding naming/function of instantiateList, but otherwise looks good to merge in.

File: minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/common/BaseSchema.java (hunk 5 above)
Comment (JPercivall) [inline minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/common/BaseSchema.java]: I agree

**Comment by JPercivall** [comment]
@apiri added change to the method name and squashed the commits, if that still looks good I can merge it or you

**Comment by apiri** [comment]
@JPercivall rock and roll 🎸  with that change.  Although I think something went awry with your imports. Travis had an issue as well, looks like the same locally.

> [INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/apiri/Development/code/apache/nifi-minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/RemoteProcessingGroupSchema.java:[57,12] error: cannot find symbol
[ERROR]   symbol:   method instantiateList(List<RemoteInputPortSchema>,String,Class<RemoteInputPortSchema>,String)
  location: class RemoteProcessingGroupSchema
/Users/apiri/Development/code/apache/nifi-minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/ConfigSchema.java:[71,12] error: cannot find symbol
[ERROR]   symbol:   method instantiateList(List<ProcessorSchema>,String,Class<ProcessorSchema>,String)
  location: class ConfigSchema
/Users/apiri/Development/code/apache/nifi-minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/ConfigSchema.java:[76,12] error: cannot find symbol
[ERROR]   symbol:   method instantiateList(List<ConnectionSchema>,String,Class<ConnectionSchema>,String)
  location: class ConfigSchema
/Users/apiri/Development/code/apache/nifi-minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/util/schema/ConfigSchema.java:[81,12] error: cannot find symbol

**Comment by apiri** [comment]
@JPercivall was waiting for Travis to do the build asynchronously but had a moment to pull down and build.  Looks good now.  Super minor, but there is a trailing quote on your commit message.  May want to scrap that before merging.

Thanks for tackling this!

**Pull request closed without merging**